### ISSUE TITLE
ApartmentAttribute and TestCaseAttribute(s) do not work together

### ISSUE THREAD

**Issue by CasperTDK** [nunit/nunit#1066]
I have tried the following:

and I get:
```csharp
System.InvalidOperationException : The calling thread must be STA, because many UI components require this.
at System.Windows.Input.InputManager..ctor()
```

I've tried to add the Apartment attribute to TestIgnore and NumericTextBoxTest as well. No changes.
Using nunit3-console.exe

```csharp
    public class NumericTextBoxTest
    {
        [TestFixture]
        public class TestIgnore
        {
            [Apartment(ApartmentState.STA)]
            [TestCase(0, Key.OemMinus, true, true, "5", false)] // does not allow negative value if text is not empty
            [TestCase(0, Key.OemMinus, true, true, "", false)] //Allows minus value if enabled
            [TestCase(0, Key.OemMinus, true, false, "", false)] //Allows minus value if decimal values not allowed
            [TestCase(0, Key.NumPad5, true, true, "", false)] //Allows numpad 5 value
            [TestCase(0, Key.D6, true, true, "", false)]//Allows d6 value
            [TestCase(1, Key.OemComma, true, true, "8", false)] //Allows decimal value if enabled and if there is a valid value
            [TestCase(1, Key.OemComma, false, true, "0", false)] //Allows decimal value if negative not allowed
            [TestCase(4, Key.OemComma, true, true, "-555", false)] //Allows decimal value if valid negative value inserted
            [TestCase(0, Key.Tab, true, true, "555", false)] //Allows decimal value if valid negative value inserted
            [TestCase(0, Key.OemComma, true, true, "", true)] //Allows decimal value without zero
            [TestCase(0, Key.OemMinus, true, true, ",5", true)] //Allows negative decimal value without zero
            public void NumericTextBoxInputTestNotIgnored(int caretIndex, Key key, bool allowNegativeValues, bool allowDecimalValues, string text, bool allowNoZeroDecimals)
            {
                    var numericTextBox = new NumericTextBox { Text = text, CaretIndex = caretIndex };

                    var isIgnored = numericTextBox.Ignore(key, allowNegativeValues, allowDecimalValues, false, false, allowNoZeroDecimals);
                    // Assert
                    Assert.IsFalse(isIgnored);
            }
        }
    }
```

**Comment by rprouse** [comment]
Does adding the apartment attribute to the test fixture work?

**Comment by CasperTDK** [comment]
Nope!

I have tried:

```javascript
 [Apartment(ApartmentState.STA)]
 public class NumericTextBoxTest
    {
```
and 
```javascript
 public class NumericTextBoxTest
    {
        [TestFixture, Apartment(ApartmentState.STA)]
        public class TestIgnore
        {
```

**Comment by CasperTDK** [comment]
I fixed the issue!

I have --Timeout=120000 as argument on my console runner.
Removing it solves the issue.

So the NumericTextBox (framework element) is not disposed before the timeout is done.
Although I wouldn't say the accompanied error make sense!

**Comment by rprouse** [comment]
I'm glad you managed to get it working. I am going to reopen this for confirmation though. When there is a timeout, we spawn a separate thread to run tests on so that we can cancel them. I want to confirm that running tests on an STA thread works with the timeout.

**Comment by CharliePoole** [comment]
@rprouse If you are checking it out, you should try it with --workers=0 as well as in normal parallel mode.

**Comment by SimonRichards** [comment]
It looks like the ApartmentAttribute is not added to Test.Properties when it is applied to a Fixture. ApartmentAttributeTests all pass through (bad) luck, however it is possible to cause failures by running this test assembly with a timeout specified on the command line.

From my limited debugging and understanding of the code, it appears that when a fixture is marked with an ApartmentAttribute, a new thread is created at WorkItem.Execute, however when the child test comes to run, TargetApartment is Unknown and we call thread.SetApartmentState(TargetApartment). It seems what we really want to do is keep the parent apartment in the case where there is no TargetApartment and we are starting a new thread (because of the non-zero timeout).

    @@ -240,7 +240,7 @@ namespace NUnit.Framework.Internal.Execution
                ApartmentState currentApartment = Thread.CurrentThread.GetApartmentState();
    
                if (Test.RequiresThread || Test is TestMethod && timeout > 0 || currentApartment != TargetApartment && TargetApartment != ApartmentState.Unknown)
    -                RunTestOnOwnThread(timeout, TargetApartment);
    +                RunTestOnOwnThread(timeout, TargetApartment == ApartmentState.Unknown ? currentApartment : TargetApartment);
                else
                    RunTest();
    #endif
    #endif

**Comment by CharliePoole** [comment]
@SimonRichards This extra thread is not always created, which may be why this bug has not been discovered. When @CasperTDK removed the --timeout option from the command-line, the extra thread was no longer needed because the fixture itself was being executed on a special queue running in the STA.

You are correct that we should not be setting the apartment state to Unknown. I think we are probably using TargetApartment == ApartmentState.Unknown to indicate that we don't care about the apartment state - not a good thing to do as it now appears. Seeing that line of code confirms the issue.

**Comment by Trass3r** [comment]
Yep same thing happened here with Apartment applied to the Fixture and Timeout attribute applied to one of the tests.

**Comment by rprouse** [comment]
The conflict between Timeout and Apartment was fixed in PR #1202. I have unit tests on a branch that show it is fixed if you use a Timeout attribute along with the Apartment attribute on a test or fixture. It also works with the `--timeout` command line option.

As @CharliePoole suspected, it is still broken with the `--workers=0` command line.
